### ISSUE TITLE
Please add mutt_buffer_log()

### ISSUE THREAD

**Issue by klartext** [neomutt/neomutt#3302]
## What would you like to be able to do?
I would like to write the content of a buffer to the logfile, and using an existing ```mutt_...()``` function for this.
I think a ```mutt_buffer_log()``` would be a good supplement to ```mutt_buffer_printf()```.

This Task could be done somehow with existing Buffer-functions and logging-functions.
But it needs some work, and  I think the task might come up frequently, when developing / debugging
functions that get a Buffer, and where you want to look at the Buffer contents close to other logging output.

Using ```mutt_buffer_printf()``` would make the output appearing after neomutt has been finished,
but not nearby the other debugging-outputs (which may have been written with ```mutt_debug()```).

The function I'm asking for may have been written by developers more than once (possibly in an imperfect way)  and being used locally, and be thrown away afterwards...

**Comment by flatcap** [comment]
It's a reasonable request, but it's a little tricky.
Behind a simple `mutt_debug()` is quite a lot of machinery.

Firstly, `mutt_debug()` and friends are macros.
They're defined to include `__FILE__`, `__LINE__` and `__func__`.
This means that the logging function knows exactly where in the code it was called from.

**Note**: This is a variadic macro; it can take any number of arguments after the log `LEVEL`.

```c
#define mutt_debug(LEVEL, ...) MuttLogger(0, __FILE__, __LINE__, __func__, LEVEL, __VA_ARGS__)
```

The macro is redirected to a global function pointer `MuttLogger`.

```c
log_dispatcher_t MuttLogger = log_disp_terminal;
```

There are multiple 'backends' for the logging, including:
- `log_disp_terminal()` -- write to the terminal
- `log_disp_curses()` -- display in the message window

We switch backends during startup from terminal to curses.


Here's the prototype for the curses backend.
```c
int log_disp_curses(time_t stamp, const char *file, int line, const char *function, enum LogLevel level, ...)
```

The log function uses `<stdarg.h>` functions to unpack the `...` arguments.
They include a `printf()`-style format string and all the arguments **it** needs.

**Comment by flatcap** [comment]
If you want something simple, like:
```c
struct Buffer *msg;
mutt_debug_buffer(LL_DEBUG1, msg);
```

Then we can probably get away with:
```c
#define mutt_debug_buffer(LEVEL, BUF)  MuttLogger(0, __FILE__, __LINE__, __func__, LL_PERROR, "%s", mutt_buffer_string(buf))
```

The `%s` protects against `%`s in the `Buffer` string.

If you want the `Buffer` as the format string, then _maybe_:
```c
#define mutt_debug_buffer(LEVEL, BUF, ...)  MuttLogger(0, __FILE__, __LINE__, __func__, LL_PERROR, mutt_buffer_string(buf), __VA_ARGS__)
```

**Comment by klartext** [comment]
The first one is a fast helper, which might make sense, even though I would like to have the possibility to provide a format string.

My workaround I used, was just a string for the name of the buffer and then the contents of the buffer plus some hard coded call to mutt_debug().
(I used mutt_buffer_string(), which you mentioned on irc.)
More flexibility with a format string would be good.

BTW: 'buf' at the replacing string should be 'BUF' I guess.

The second one - buffer as format string (?!) seems a bit too much, I think.
Would there be a reason for such a beast? ;-)

BTW: while trying out the first macro in the hook-parser refactoring, there suddenly appeared an error messages in the ```s``` buffer (comes as argument). Maybe a problem with the global logging pointer you mentioned above?
(Removing globals might be worth some effort. in general in neomutt...)

**Comment by flatcap** [comment]
I've just fixed a few typos

**Comment by flatcap** [comment]
> suddenly appeared an error messages

That's because I accidentally cut'n'pasted some **error** log levels.
I've fixed the param to `LEVEL`, now.

**Comment by klartext** [comment]
Ah, ok :-)

**Comment by klartext** [comment]
A "\n" at the end would also not be too bad... but maybe format strings would be the best option.

**Comment by flatcap** [comment]
> (Removing globals might be worth some effort. in general in neomutt...)

Yes and no.  There are still lots to eliminate, but `MuttLogger` needs to be global.

> A "\n" at the end would also not be too bad

That's historic.

The "visual" levels: message, warning and erorr, don't have `\n`
The debug levels **do**.  They used to only be flushed when `\n` was received.
Changing them now would be quite a large job (massive understatement:-)

**Comment by flatcap** [comment]
Sample code

```c
  struct Buffer buf = mutt_buffer_make(256);
  mutt_buffer_strcpy(&buf, "hello %d world\n");
  mutt_debug_buffer(LL_DEBUG1, &buf, 42);
  mutt_buffer_dealloc(&buf);
```

**Comment by klartext** [comment]
> > (Removing globals might be worth some effort. in general in neomutt...)
> 
> Yes and no. There are still lots to eliminate, but `MuttLogger` needs to be global.

Yes, after later rethinking what I have written before, I also came to the conclusion, that for the logger the (one) global might not be that bad. One global pointer, not thousands of global variables.

**Comment by klartext** [comment]
> Sample code
> 
> ```c
>   struct Buffer buf = mutt_buffer_make(256);
>   mutt_buffer_strcpy(&buf, "hello %d world\n");
>   mutt_debug_buffer(LL_DEBUG1, &buf, 42);
>   mutt_buffer_dealloc(&buf);
> ```

OK, nice. Thanks for the example.
...might be called ```mutt_debug_buffer2``` ?

Will explore that stuff in more depth soon. Enough coding for now.

**Comment by gahr** [comment]
I have read this issue and the backlog on IRC, but I fail to see any advantage over `mutt_log(LL_DEBUG1, mutt_buffer_string(&buf), ...)`
